core/src/lib.rs

```diff
@@ -27,3 +27,9 @@ pub mod util;
 pub use primary_keys::MongoPrimaryKeys;
 mod foreign_keys;
 pub use foreign_keys::MongoForeignKeys;
+
+#[cfg(target_os = "macos")]
+pub type WChar = u32;
+
+#[cfg(not(target_os = "macos"))]
+pub type WChar = u16;
```

Comment:
Without some plumbing changes, this will cause many headaches developing on macos. Running `cargo clippy` on my mac shows several errors:

```
error[E0308]: mismatched types
  --> integration_test/src/../tests/common/mod.rs:56:13
   |
55 |         String::from_utf16_lossy(slice::from_raw_parts(
   |                                  --------------------- arguments to this function are incorrect
56 |             actual_message_text as *const WChar,
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u16`, found `u32`
   |
   = note: expected raw pointer `*const u16`
              found raw pointer `*const u32`
note: function defined here
  --> /Users/nathanleniz/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/slice/raw.rs:92:21
   |
92 | pub const unsafe fn from_raw_parts<'a, T>(data: *const T, len: usize) -> &'a [T] {
   |                     ^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
error: could not compile `integration_test` due to previous error
warning: build failed, waiting for other jobs to finish...
error[E0308]: mismatched types
    --> odbc/src/api/diag.rs:40:45
     |
40   |     copy_nonoverlapping(state_u16.as_ptr(), output_ptr, 6);
     |     -------------------                     ^^^^^^^^^^ expected `u16`, found `u32`
     |     |
     |     arguments to this function are incorrect
     |
     = note: expected raw pointer `*mut u16`
                found raw pointer `*mut u32`
note: function defined here
    --> /Users/nathanleniz/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:2346:21
     |
2346 | pub const unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize) {
     |                     ^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> odbc/src/api/data.rs:969:41
    |
969 |         return String::from_utf16_lossy(&dst);
    |                ------------------------ ^^^^ expected slice `[u16]`, found struct `std::vec::Vec`
    |                |
    |                arguments to this function are incorrect
    |
    = note: expected reference `&[u16]`
               found reference `&std::vec::Vec<u32>`
note: associated function defined here
   --> /Users/nathanleniz/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/string.rs:724:12
    |
724 |     pub fn from_utf16_lossy(v: &[u16]) -> String {
    |            ^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> odbc/src/api/data.rs:975:30
    |
975 |     String::from_utf16_lossy(&dst)
    |     ------------------------ ^^^^ expected slice `[u16]`, found struct `std::vec::Vec`
    |     |
    |     arguments to this function are incorrect
    |
    = note: expected reference `&[u16]`
               found reference `&std::vec::Vec<u32>`
note: associated function defined here
   --> /Users/nathanleniz/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/string.rs:724:12
    |
724 |     pub fn from_utf16_lossy(v: &[u16]) -> String {
    |            ^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
    --> odbc/src/api/data.rs:1003:43
     |
1003 |     copy_nonoverlapping(message.as_ptr(), output_ptr, num_chars - 1);
     |     -------------------                   ^^^^^^^^^^ expected `u16`, found `u32`
     |     |
     |     arguments to this function are incorrect
     |
     = note: expected raw pointer `*mut u16`
                found raw pointer `*mut u32`
note: function defined here
    --> /Users/nathanleniz/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs:2346:21
     |
2346 | pub const unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize) {
     |                     ^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
    --> odbc/src/api/data.rs:1004:38
     |
1004 |     *output_ptr.add(num_chars - 1) = 0u16;
     |     ------------------------------   ^^^^ expected `u32`, found `u16`
     |     |
     |     expected due to the type of this binding

error: could not compile `mongo-odbc-driver` due to 5 previous errors
```

I propose we separate out refactoring to support macos in another ticket since it's not necessary to accomplish our immediate goal. WDYT? cc @nbagnard

Reply:
I think I just need to use utf-32 there on mac.

Reply:
Ok, I think I have a good solution to this up. Need to update the tests so they will pass on macos (as a proxy for utf32) still, though.